pkg/sbom/diff_nodelist.go

```diff
@@ -0,0 +1,256 @@
+package sbom
+
+import (
+	"regexp"
+	"sort"
+	"strings"
+)
+
+// NodeListDiff represents the difference between two lists NodeLists.
+type NodeListDiff struct {
+	NodesDiff         NodeListDiffNodes
+	EdgesDiff         NodeListDiffEdges
+	RootElmementsDiff NodeListRootElementDiff
+}
+
+func (nlDiff *NodeListDiff) AddedNodeList() *NodeList {
+	var addedNodeList NodeList
+	for _, addedNode := range nlDiff.NodesDiff.Added {
+		addedNodeList.AddNode(addedNode)
+	}
+	for _, addedEdge := range nlDiff.EdgesDiff.Added {
+		addedNodeList.AddEdge(addedEdge)
+	}
+	addedNodeList.RootElements = append(addedNodeList.RootElements, nlDiff.RootElmementsDiff.Added...)
+
+	for _, diffNodeAdded := range nlDiff.NodesDiff.NodeDiff {
+		addedNodeList.AddNode(diffNodeAdded.Added)
+	}
+
+	return &addedNodeList
+}
+
+func (nlDiff *NodeListDiff) RemovedNodeList() *NodeList {
+	var removedNodeList NodeList
+	for _, removedNode := range nlDiff.NodesDiff.Removed {
+		removedNodeList.AddNode(removedNode)
+	}
+	for _, removedEdge := range nlDiff.EdgesDiff.Removed {
+		removedNodeList.AddEdge(removedEdge)
+	}
+	removedNodeList.RootElements = append(removedNodeList.RootElements, nlDiff.RootElmementsDiff.Removed...)
+
+	for _, diffNodeRemoved := range nlDiff.NodesDiff.NodeDiff {
+		removedNodeList.AddNode(diffNodeRemoved.Removed)
+	}
+
+	return &removedNodeList
+}
+
+// NodeListDiffNodes represents the differences between two NodeList nodes.
+type NodeListDiffNodes struct {
+	Added    []*Node
+	Removed  []*Node
+	NodeDiff []*NodeDiff
```

Comment:
This data structure seems problematic to me. As the NodeDiff array gets populated, the original context of the modified node gets lost. In other words, we can know that some nodes had changes but we don't know which.

Reply:
Agreed,
Maybe we encapsulate the original Documents in the result as and somehow reference the original node.

I am open to your suggestion though.

Reply:
I think this can be solved by extending the NodeDiff to capture the original node's ID. The structure would end up something like:

```golang 
type NodeDiff struct {
	NodeID string
	Added     *Node
	Removed   *Node
	DiffCount int
}
```
Since we are looking at different scenarios where we may be diffing nodes with different IDs (such as the case you had, diffing nodes with related purls ). For those cases, we may want to ensure we are capturing the ID from nl2 in the `NodeDiff.Added` result.

Reply:
Ok Cool, So we end up just adding a NodeID to the NodeDiff? 
As For retaining the Node ID i think that because the diff operates on the Node.ID as a index it should always be present.